File: tests/data_structures/binary_tree/binary_tree.py
"""
Binary Tree benchmark implementation in Python.
Tests basic binary search tree operations: insert, search, traverse.
"""

import time


class TreeNode:
    """Node for binary tree."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinarySearchTree:
    """Binary search tree implementation."""
    
    def __init__(self):
        self.root = None
        self.size = 0
    
    def insert(self, val):
        """Insert a value into the tree."""
        if self.root is None:
            self.root = TreeNode(val)
            self.size += 1
        else:
            self._insert_recursive(self.root, val)
    
    def _insert_recursive(self, node, val):
        """Recursively insert a value."""
        if val < node.val:
            if node.left is None:
                node.left = TreeNode(val)
                self.size += 1
            else:
                self._insert_recursive(node.left, val)
        elif val > node.val:
            if node.right is None:
                node.right = TreeNode(val)
                self.size += 1
            else:
                self._insert_recursive(node.right, val)
        # Equal values are ignored (no duplicates)
    
    def search(self, val):
        """Search for a value in the tree."""
        return self._search_recursive(self.root, val)
    
    def _search_recursive(self, node, val):
        """Recursively search for a value."""
        if node is None:
            return False
        if node.val == val:
            return True
        elif val < node.val:
            return self._search_recursive(node.left, val)
        else:
            return self._search_recursive(node.right, val)
    
    def inorder_traversal(self):
        """Perform inorder traversal."""
        result = []
        self._inorder_recursive(self.root, result)
        return result
    
    def _inorder_recursive(self, node, result):
        """Recursively perform inorder traversal."""
        if node is not None:
            self._inorder_recursive(node.left, result)
            result.append(node.val)
            self._inorder_recursive(node.right, result)
    
    def get_size(self):
        """Get the size of the tree."""
        return self.size


def main():
    """Run binary tree benchmark."""
    print("Starting binary tree benchmark...")
    start_time = time.perf_counter()
    
    # Create binary search tree
    bst = BinarySearchTree()
    
    # Insert operations with a larger dataset
    nodes_count = 1000
    import random
    random.seed(42)  # For reproducible results
    values = list(range(nodes_count))
    random.shuffle(values)
    
    for val in values:
        bst.insert(val)
    
    # Search operations
    found_count = 0
    search_values = random.sample(values, min(100, len(values)))
    for val in search_values:
        if bst.search(val):
            found_count += 1
    
    # Traversal operation
    traversal_result = bst.inorder_traversal()
    is_sorted = all(traversal_result[i] <= traversal_result[i + 1] 
                   for i in range(len(traversal_result) - 1))
    
    end_time = time.perf_counter()
    execution_time = end_time - start_time
    
    print(f"Tree operations completed: {nodes_count} inserts, {found_count} searches")
    print(f"Final tree size: {bst.get_size()}")
    print(f"Inorder traversal length: {len(traversal_result)}")
    print(f"Traversal is sorted: {is_sorted}")
    print(f"Execution time: {execution_time:.6f} seconds")


if __name__ == "__main__":
    main()

File: tests/data_structures/binary_tree/binary_tree.ts
/**
 * Binary Tree benchmark implementation in TypeScript.
 * Tests basic binary search tree operations: insert, search, traverse.
 */

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    
    constructor(val: number) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    private root: TreeNode | null;
    private size: number;
    
    constructor() {
        this.root = null;
        this.size = 0;
    }
    
    insert(val: number): void {
        if (this.root === null) {
            this.root = new TreeNode(val);
            this.size++;
        } else {
            this.insertRecursive(this.root, val);
        }
    }
    
    private insertRecursive(node: TreeNode, val: number): void {
        if (val < node.val) {
            if (node.left === null) {
                node.left = new TreeNode(val);
                this.size++;
            } else {
                this.insertRecursive(node.left, val);
            }
        } else if (val > node.val) {
            if (node.right === null) {
                node.right = new TreeNode(val);
                this.size++;
            } else {
                this.insertRecursive(node.right, val);
            }
        }
        // Equal values are ignored (no duplicates)
    }
    
    search(val: number): boolean {
        return this.searchRecursive(this.root, val);
    }
    
    private searchRecursive(node: TreeNode | null, val: number): boolean {
        if (node === null) {
            return false;
        }
        if (node.val === val) {
            return true;
        } else if (val < node.val) {
            return this.searchRecursive(node.left, val);
        } else {
            return this.searchRecursive(node.right, val);
        }
    }
    
    inorderTraversal(): number[] {
        const result: number[] = [];
        this.inorderRecursive(this.root, result);
        return result;
    }
    
    private inorderRecursive(node: TreeNode | null, result: number[]): void {
        if (node !== null) {
            this.inorderRecursive(node.left, result);
            result.push(node.val);
            this.inorderRecursive(node.right, result);
        }
    }
    
    getSize(): number {
        return this.size;
    }
}

// Simple pseudo-random number generator for reproducible results
class SimpleRandom {
    private seed: number;
    
    constructor(seed: number) {
        this.seed = seed;
    }
    
    next(): number {
        this.seed = (this.seed * 1103515245 + 12345) & 0x7fffffff;
        return this.seed / 0x7fffffff;
    }
    
    shuffle<T>(array: T[]): void {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(this.next() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
    }
}

function isSorted(arr: number[]): boolean {
    for (let i = 1; i < arr.length; i++) {
        if (arr[i - 1] > arr[i]) {
            return false;
        }
    }
    return true;
}

function main(): void {
    console.log("Starting binary tree benchmark...");
    const startTime = process.hrtime.bigint();
    
    const bst = new BinarySearchTree();
    const nodesCount = 1000;
    
    // Create shuffled values for insertion
    const rng = new SimpleRandom(42);
    const values: number[] = [];
    for (let i = 0; i < nodesCount; i++) {
        values.push(i);
    }
    rng.shuffle(values);
    
    // Insert operations
    for (const val of values) {
        bst.insert(val);
    }
    
    // Search operations
    let foundCount = 0;
    const searchCount = Math.min(100, values.length);
    for (let i = 0; i < searchCount; i++) {
        if (bst.search(values[i])) {
            foundCount++;
        }
    }
    
    // Traversal operation
    const traversalResult = bst.inorderTraversal();
    const sorted = isSorted(traversalResult);
    
    const endTime = process.hrtime.bigint();
    const executionTime = Number(endTime - startTime) / 1e9; // Convert to seconds
    
    console.log(`Tree operations completed: ${nodesCount} inserts, ${foundCount} searches`);
    console.log(`Final tree size: ${bst.getSize()}`);
    console.log(`Inorder traversal length: ${traversalResult.length}`);
    console.log(`Traversal is sorted: ${sorted}`);
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
}

main();